Implement model refreshing for Teams tab

andy-armstrong opened the pull request:
https://openedx.atlassian.net/browse/TNL-2976

This is (hopefully) a comprehensive fix for all the refresh problems that we were having with teams. The basic approach is that our three collections (topics, teams, and team memberships) all now inherit from a base collection which tracks whether the collection should be considered stale. It is each view's job to notice if the models that it is rendering are stale, and if so to first refetch them from the server. Hopefully this ensures that all the views are showing up-to-date information without requiring that every collection be updated on every change.

andy-armstrong commented:
@cahrens @peter-fogg Please review.

andy-armstrong commented:
FYI @symbolist @dan-f @bderusha

peter-fogg commented on common/static/common/js/spec_helpers/ajax_helpers.js:
```diff
@@ -1,5 +1,5 @@
 define(['sinon', 'underscore'], function(sinon, _) {
-    var fakeServer, fakeRequests, expectRequest, expectJsonRequest, expectPostRequest,
+    var fakeServer, fakeRequests, expectRequest, expectJsonRequest, expectPostRequest, getCanonicalUrl,
```
Looks like `getCanonicalUrl` isn't used.

cahrens commented:
Those failing JS tests were removed by Diana-- rebase issues.

cahrens commented:
Jenkins run bokchoy

peter-fogg commented on lms/djangoapps/teams/static/teams/js/spec/views/team_profile_spec.js:
```diff
@@ -145,29 +167,76 @@ define([
+                        },
+                            {
+                                'user': {
+                                    'username': 'bilbo1',
```
Not Frodo?

peter-fogg commented on lms/djangoapps/teams/static/teams/js/spec/views/team_profile_spec.js:
```diff
@@ -145,29 +167,76 @@ define([
+                    var requests = AjaxHelpers.requests(this);
 
+                    var view = createTeamProfileView(
+                        requests, { country: 'US', language: 'en', membership: DEFAULT_MEMBERSHIP}
```
This is a stupid nit that shouldn't block this from merging, but above there's no space between `{` and `country`.

cahrens commented on lms/djangoapps/teams/static/teams/js/collections/team.js:
```diff
@@ -1,18 +1,22 @@
+                        },
+                        BaseCollection.prototype.server_api
+                    );
+                    delete this.server_api.sort_order; // Sort order is not specified for the Team API
```
I'm confused about sort order and order_by. Not related to your changes though....

peter-fogg commented on lms/djangoapps/teams/static/teams/js/views/my_teams.js:
```diff
@@ -5,9 +5,19 @@
+                                self.render();
+                            });
+                    } else {
+                        TeamsView.prototype.render.call(this);
```
Why are we only calling the prototype's render if the collection isn't stale? Does this get called eventually in the stale case?

andy-armstrong replied:
Yes, it calls ``self.render()`` once it successfully fetches the collection.

cahrens commented:
jenkins run bokchoy

peter-fogg commented on lms/djangoapps/teams/static/teams/js/views/topic_teams.js:
```diff
@@ -17,19 +17,40 @@
-                        this.$el.append(_.template(teamActionsTemplate, {message: message}));
+                    var self = this,
+                        collectionRefresh;
+                    if (this.collection.isStale || this.teamMemberships.isStale) {
```
It might be nice to factor the code that happens if `isStale` into separate methods.

andy-armstrong commented:
I've fixed the bad rebase and rebased to pick up all the subsequent changes to master. I'll see how this build goes and then will address the feedback.

cahrens commented:
We still need to sort out the bok choy test failure-- according to the screenshot, the user did leave the team. The refresh must have changed something about what it is waiting for. Hopefully it is reproducible on devstack-- I'll take a look as soon as I get in this morning.

cahrens commented on lms/djangoapps/teams/static/teams/js/collections/base.js:
```diff
@@ -0,0 +1,23 @@
+
+                onUpdate: function(event) {
+                    this.isStale = true;
+                }
```
I was hoping that a helper method to do this would also be in this base class:

if (this.collection.isStale) {
+                    this.collection.fetch()
+                        .done(function() {
+                            self.collection.isStale = false;
+                            self.render();
+                        });
+                }

If you don't think it makes sense, leave it as-is.

andy-armstrong replied:
I'm conflicted because the ``self`` here is the view, not the collection. How about we have a ``refresh`` method on these collections that does the fetch and clears the stale flag, and then the view call:

    self.collection.refresh().done(...)

I'll try that and see how it feels.

cahrens commented on lms/djangoapps/teams/static/teams/js/spec/views/my_teams_spec.js:
```diff
@@ -28,26 +30,52 @@ define([
+                request, url, parameters;
+            TeamSpecHelpers.verifyCards(myTeamsView, []);
+            expect(myTeamsView.$el.text().trim()).toBe('You are not currently a member of any team.');
+            teamMemberships.teamEvents.trigger('teams:update', { action: 'create' });
```
It does the same for add and join team, correct? Not that we necessarily need to test them, just want to make sure.

andy-armstrong replied:
This particular case works for all updates and ignores the action, but it is probably worth checking all three.

cahrens commented on lms/djangoapps/teams/static/teams/js/views/team_profile.js:
```diff
@@ -73,10 +73,18 @@
+                        });
+                        view.model.fetch({})
+                            .done(function() {
+                                view.model.isStale = false;
```
I'm wondering if the teamEvents trigger should happen in this "done" since the teams:update sends view.model as the team (and it won't be up-to-date until the fetch is complete).

cahrens commented on lms/djangoapps/teams/static/teams/js/views/team_join.js:
```diff
@@ -1,108 +1,115 @@
+                            action: 'join',
+                            team: view.model
+                        });
+                        view.model.fetch()
```
I'm wondering if the teamEvents trigger should happen in this "done" since the teams:update sends view.model as the team (and it won't be up-to-date until the fetch is complete).

cahrens commented on lms/djangoapps/teams/static/teams/js/views/topic_teams.js:
```diff
@@ -17,19 +17,40 @@
+                                        self.collection.isStale = false;
+                                    });
+                        } else {
+                            collectionRefresh = $.Deferred().resolve().promise();
```
I"m having trouble following this deferred logic. Perhaps you can explain it to me this morning.

andy-armstrong replied:
You can tell I wrote this too late at night. The new version I have is *much* clearer. :-)

cahrens commented:
I noted a few things, otherwise :+1:. Although it being the morning with a whole 4 hours before the release is cut, I would like a bit more Jasmine test coverage of the various views that respond to events. If that's not ready for the release, perhaps merge another PR after the release?

peter-fogg commented:
:+1:

cahrens commented:
:+1:

cahrens commented:
Lettuce tests all passed, they failed to record-- will report in testeng.
